chapter01/src/main/java/com/seaofnodes/simple/parser/AST.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * DO NOT REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ * <p>
+ * Contributor(s):
+ * <p>
+ * The Initial Developer of the Original Software is Dibyendu Majumdar.
+ * <p>
+ * This file is part of the Sea Of Nodes Simple Programming language
+ * implementation. See https://github.com/SeaOfNodes/Simple
+ * <p>
+ * The contents of this file are subject to the terms of the
+ * Apache License Version 2 (the "APL"). You may not use this
+ * file except in compliance with the License. A copy of the
+ * APL may be obtained from:
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package com.seaofnodes.simple.parser;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A simple AST definition.
+ */
+public class AST {
+
+    protected AST() {
+    }
+
+
+    public abstract static class Expr extends AST {
+        public final Token op;
+
+        protected Expr(Token op) {
+            this.op = op;
+        }
+    }
+
+    /**
+     * A name in the language
+     */
+    public static class Identifier extends Expr {
+        public final String name;
+
+        public Identifier(Token op) {
+            super(op);
+            name = op.str;
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+    }
+
+    /**
+     * Reference to a name in the language
+     */
+    public static class Symbol extends Expr {
+        public final String name;
+
+        public Symbol(Token op) {
+            super(op);
+            name = op.str;
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+    }
+
+    public static class Binary extends Expr {
+        public final Expr expr1;
+        public final Expr expr2;
+
+        public Binary(Token op, Expr expr1, Expr expr2) {
+            super(op);
+            this.expr1 = expr1;
+            this.expr2 = expr2;
+        }
+
+        @Override
+        public String toString() {
+            return "(" + expr1 + op.toString() + expr2 + ")";
```

Comment:
Since `op` has a toString already, why not `"(" + kids[0]+ op + kids[1] + ")"` OR
`"(" + lhs()+ op + rhs() + ")"`